docs/architecture.md

```diff
@@ -0,0 +1,473 @@
+    // Await the next transaction from network gossip (similar to stream.next()).
+    async fn next_gossiped_transaction(&self) -> Self::GossipedTransaction;
+    // Report the validity of a transaction received from the network.
+    fn notify_gossip_transaction_validity(message: &Self::GossipedTransaction, validity: GossipValidity);
+}
+
+// Generic wrapper for data received from the network
+pub trait NetworkData<D> {
+    // Transfer ownership of the provided data without consuming any message metadata
+    fn take_data(&mut self) -> Option<D>;
+}
+
+pub enum GossipValidity {
+    // Report whether the gossiped message is valid and safe to rebroadcast
+    Accept,
+    // Ignore the received message and prevent further gossiping
+    Ignore,
+    // Punish the gossip sender for providing invalid 
+    // (or malicious) data and prevent further gossiping
+    Invalid,
+}
+
+pub trait Database:
+    StorageInspect<Coins, Error = KvStoreError>
+    + StorageInspect<ContractsRawCode, Error = KvStoreError>
+    + StorageInspect<Messages, Error = KvStoreError>
+    + Send
+    + Sync
+{}
+```
+
+#### Adapters: fuel_core::service::adapters::transaction_pool
+```rust
+impl block_importer::ports::TransactionPool for Service<TransactionPool> {
+    // insert impl here
+}
+
+impl block_producer::ports::TransactionPool for Service<TransactionPool> {
+    // insert impl here
+}
+```
+
+### Executor
+
+#### Ports: fuel_block_executor::ports
+```rust
+trait Database: IntepreterStorage {
+    fn get_coin(id: UtxoId) -> Result<Coin>;
+    fn upsert_coin(coin: Coin) -> Result<()>;
+    fn get_message(message_id: MessageId) -> Result<Message>;
+    fn mark_message_spent(message_id: MessageId) -> Result<()>;
+    fn get_contract_utxo_id(contract_id: ContractId) -> Result<UtxoId>;
+    fn set_contract_utxo_id(contract_id: ContractId, utxo_id: UtxoId) -> Result<()>;
+    fn get_block(height: BlockHeight) -> Result<FuelBlock>;
+    fn insert_block(block: FuelBlock) -> Result<()>;
+    // start a nested transaction to allow changes to be rolled-back
+    fn transaction(&self) -> Box<dyn DatabaseTransaction>;
+}
+
+// Database overlay that accumulates changes until committed at the end.
+trait DatabaseTransaction: AsRef<Box<dyn Database>> {
+    // consume the transaction and commit the changes
+    fn commit(self) -> Result<()>;
+}
+```
+#### Adapters: fuel_core::service::adapters::executor
+```rust
+impl fuel_block_producer::ports::Executor for Service<Executor> {
+    // insert impl here
+}
+
+impl fuel_block_importer::ports::Executor for Service<Executor> {
+    // insert impl here
+}
+```
+
+### Relayer
+
+#### Ports: fuel_relayer::ports
+```rust
+trait Database: StorageMutate<Messages, Error = KvStoreError> {
+    fn get_current_sealed_block(&self) -> Result<SealedFuelBlock>;
+    fn get_da_synced_height(&self) -> Result<DaBlockHeight>;
+    fn set_da_synced_height(&self, height: DaBlockHeight) -> Result<Option<DaBlockHeight>>;
```

Comment:
What does the return type mean here?

Reply:
maybe previously saved value? probably not needed